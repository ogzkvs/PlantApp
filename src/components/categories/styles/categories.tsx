import {
  StyleSheet,
  ViewStyle,
  ImageStyle,
  TextStyle,
  Dimensions,
} from 'react-native';

const { width } = Dimensions.get('window');

const padding = 30;
const imageWidth = width / 2 - padding;
const imageHeight = 152;
const aspectRatio = imageWidth / imageHeight;

interface Styles {
  container: ViewStyle;
  image: ImageStyle;
  title: TextStyle;
  textContainer: ViewStyle;
  listContent: ViewStyle;
  itemContainer: ViewStyle;
}

export default StyleSheet.create<Styles>({
  container: {
    flex: 1,
  },
  itemContainer: {
    flex: 1,
    marginBottom: 15,
  },
  image: {
    width: imageWidth,
    aspectRatio: aspectRatio,
  },
  textContainer: {
    padding: 14,
  },
  title: {
    color: '#13231B',
    fontSize: 16,
    lineHeight: 21,
  },
  listContent: {
    marginVertical: 10,
    flex: 1,
  },
});
