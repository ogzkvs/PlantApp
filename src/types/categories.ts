export interface CategoryResponse {
  data: CategoryType[];
  meta: Meta;
}

export interface CategoryType {
  id: number;
  name: string;
  createdAt: string;
  updatedAt: string;
  publishedAt: string;
  title: string;
  rank: number;
  image: Image;
}

export interface Image {
  id: number;
  name: string;
  alternativeText: any;
  caption: any;
  width: number;
  height: number;
  formats: any;
  hash: string;
  ext: string;
  mime: string;
  size: number;
  url: string;
  previewUrl: any;
  provider: string;
  provider_metadata: any;
  createdAt: string;
  updatedAt: string;
}

export interface Meta {
  pagination: Pagination;
}

export interface Pagination {
  page: number;
  pageSize: number;
  pageCount: number;
  total: number;
}

export interface CategoriesState {
  data: CategoryType[];
  loading: boolean;
  error: string | null;
}
