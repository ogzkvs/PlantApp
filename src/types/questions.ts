export interface QuestionType {
  id: number;
  title: string;
  subtitle: string;
  image_uri: string;
  uri: string;
  order: number;
}

export interface QuestionsState {
  data: QuestionType[];
  loading: boolean;
  error: string | null;
}
