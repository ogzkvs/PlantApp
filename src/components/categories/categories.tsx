import React from 'react';
import { View, FlatList } from 'react-native';
import { ImageBackground } from 'react-native';
import CustomText from '../customText';
import { CategoryType } from '../../types/categories';
import styles from './styles/categories';

type CategoriesProps = {
  data: CategoryType[];
};

const Categories: React.FC<CategoriesProps> = ({ data }) => {
  const renderItem = ({
    item,
    index,
  }: {
    item: CategoryType;
    index: number;
  }) => {
    return (
      <View
        style={[{ marginLeft: index % 2 !== 0 ? 15 : 0 }, styles.itemContainer]}
      >
        <ImageBackground
          source={{ uri: item?.image?.url }}
          style={styles.image}
          imageStyle={{
            borderRadius: 12,
            borderWidth: 0.5,
            borderColor: '#29BB892E',
          }}
        >
          <View style={styles.textContainer}>
            <CustomText fontFamily="Rubik-Medium" style={styles.title}>
              {item?.title}
            </CustomText>
          </View>
        </ImageBackground>
      </View>
    );
  };

  return (
    <FlatList
      scrollEnabled={false}
      numColumns={2}
      data={data}
      showsHorizontalScrollIndicator={false}
      keyExtractor={item => `categories-${item.id}`}
      renderItem={renderItem}
      contentContainerStyle={styles.listContent}
    />
  );
};

export default Categories;
