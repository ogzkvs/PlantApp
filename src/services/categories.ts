import { API_BASE_URL } from '../config';
import { CategoryResponse, CategoryType } from '../types/categories';

export class CategoryService {
  static async fetchCategories(): Promise<CategoryType[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/getCategories`);
      const json: CategoryResponse = await response.json();

      return json?.data as CategoryType[];
    } catch (error) {
      console.error('Failed to fetch categories:', error);
      throw error;
    }
  }
}
