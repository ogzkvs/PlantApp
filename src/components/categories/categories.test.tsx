import React from 'react';
import { render } from '@testing-library/react-native';
import Categories from './Categories';
import { CategoryType } from '../../types/categories';
import { fireEvent } from '@testing-library/react-native';

const mockData: CategoryType[] = [
  {
    id: 11,
    name: 'fern',
    createdAt: '2023-01-11T10:53:05.801Z',
    updatedAt: '2023-01-11T10:54:30.059Z',
    publishedAt: '2023-01-11T10:53:07.416Z',
    title: 'Ferns',
    rank: 0,
    image: {
      id: 23,
      name: '6.png',
      alternativeText: null,
      caption: null,
      width: 116,
      height: 126,
      formats: null,
      hash: '6_edbcc6988a',
      ext: '.png',
      mime: 'image/png',
      size: 8.24,
      url: 'https://cms-cdn.plantapp.app/6_edbcc6988a/6_edbcc6988a.png',
      previewUrl: null,
      provider: '@strapi-community/strapi-provider-upload-google-cloud-storage',
      provider_metadata: null,
      createdAt: '2023-01-11T10:44:46.151Z',
      updatedAt: '2023-01-11T10:44:46.151Z',
    },
  },
  {
    id: 10,
    name: 'cacti-and-succulent',
    createdAt: '2023-01-11T10:52:28.521Z',
    updatedAt: '2023-01-11T10:54:39.391Z',
    publishedAt: '2023-01-11T10:52:36.428Z',
    title: 'Cacti and Succulents',
    rank: 1,
    image: {
      id: 25,
      name: '5.png',
      alternativeText: null,
      caption: null,
      width: 158,
      height: 152,
      formats: null,
      hash: '5_d2384a3938',
      ext: '.png',
      mime: 'image/png',
      size: 10.01,
      url: 'https://cms-cdn.plantapp.app/5_d2384a3938/5_d2384a3938.png',
      previewUrl: null,
      provider: '@strapi-community/strapi-provider-upload-google-cloud-storage',
      provider_metadata: null,
      createdAt: '2023-01-11T10:50:17.828Z',
      updatedAt: '2023-01-11T10:51:05.935Z',
    },
  },
];

describe('Categories Component', () => {
  it('renders all category titles correctly', () => {
    const { getByText } = render(<Categories data={mockData} />);

    expect(getByText('Ferns')).toBeTruthy();
    expect(getByText('Cacti and Succulents')).toBeTruthy();
  });

  it('renders correct number of items', () => {
    const { getAllByText } = render(<Categories data={mockData} />);

    const titles = mockData.map(c => c.title);
    titles.forEach(title => {
      expect(getAllByText(title).length).toBeGreaterThan(0);
    });
  });

  it('renders correct images with correct URLs', () => {
    const { getAllByTestId } = render(<Categories data={mockData} />);
    const images = getAllByTestId('category-image');

    expect(images.length).toBe(mockData.length);

    images.forEach((image, index) => {
      expect(image.props.source.uri).toBe(mockData[index].image.url);
    });
  });

  it('matches snapshot', () => {
    const tree = render(<Categories data={mockData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
