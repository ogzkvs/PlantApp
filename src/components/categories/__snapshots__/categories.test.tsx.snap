// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories Component matches snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "flex": 1,
      "marginVertical": 10,
    }
  }
  data={
    [
      {
        "createdAt": "2023-01-11T10:53:05.801Z",
        "id": 11,
        "image": {
          "alternativeText": null,
          "caption": null,
          "createdAt": "2023-01-11T10:44:46.151Z",
          "ext": ".png",
          "formats": null,
          "hash": "6_edbcc6988a",
          "height": 126,
          "id": 23,
          "mime": "image/png",
          "name": "6.png",
          "previewUrl": null,
          "provider": "@strapi-community/strapi-provider-upload-google-cloud-storage",
          "provider_metadata": null,
          "size": 8.24,
          "updatedAt": "2023-01-11T10:44:46.151Z",
          "url": "https://cms-cdn.plantapp.app/6_edbcc6988a/6_edbcc6988a.png",
          "width": 116,
        },
        "name": "fern",
        "publishedAt": "2023-01-11T10:53:07.416Z",
        "rank": 0,
        "title": "Ferns",
        "updatedAt": "2023-01-11T10:54:30.059Z",
      },
      {
        "createdAt": "2023-01-11T10:52:28.521Z",
        "id": 10,
        "image": {
          "alternativeText": null,
          "caption": null,
          "createdAt": "2023-01-11T10:50:17.828Z",
          "ext": ".png",
          "formats": null,
          "hash": "5_d2384a3938",
          "height": 152,
          "id": 25,
          "mime": "image/png",
          "name": "5.png",
          "previewUrl": null,
          "provider": "@strapi-community/strapi-provider-upload-google-cloud-storage",
          "provider_metadata": null,
          "size": 10.01,
          "updatedAt": "2023-01-11T10:51:05.935Z",
          "url": "https://cms-cdn.plantapp.app/5_d2384a3938/5_d2384a3938.png",
          "width": 158,
        },
        "name": "cacti-and-succulent",
        "publishedAt": "2023-01-11T10:52:36.428Z",
        "rank": 1,
        "title": "Cacti and Succulents",
        "updatedAt": "2023-01-11T10:54:39.391Z",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEnabled={false}
  scrollEventThrottle={0.0001}
  showsHorizontalScrollIndicator={false}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            [
              {
                "marginLeft": 0,
              },
              {
                "flex": 1,
                "marginBottom": 15,
              },
            ]
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              {
                "aspectRatio": 2.2697368421052633,
                "width": 345,
              }
            }
          >
            <Image
              source={
                {
                  "uri": "https://cms-cdn.plantapp.app/6_edbcc6988a/6_edbcc6988a.png",
                }
              }
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  {
                    "height": undefined,
                    "width": 345,
                  },
                  {
                    "borderColor": "#29BB892E",
                    "borderRadius": 12,
                    "borderWidth": 0.5,
                  },
                ]
              }
              testID="category-image"
            />
            <View
              style={
                {
                  "padding": 14,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Rubik-Medium",
                    },
                    {
                      "color": "#13231B",
                      "fontSize": 16,
                      "lineHeight": 21,
                    },
                  ]
                }
                testID="category-title"
              >
                Ferns
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "marginLeft": 15,
              },
              {
                "flex": 1,
                "marginBottom": 15,
              },
            ]
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              {
                "aspectRatio": 2.2697368421052633,
                "width": 345,
              }
            }
          >
            <Image
              source={
                {
                  "uri": "https://cms-cdn.plantapp.app/5_d2384a3938/5_d2384a3938.png",
                }
              }
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  {
                    "height": undefined,
                    "width": 345,
                  },
                  {
                    "borderColor": "#29BB892E",
                    "borderRadius": 12,
                    "borderWidth": 0.5,
                  },
                ]
              }
              testID="category-image"
            />
            <View
              style={
                {
                  "padding": 14,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Rubik-Medium",
                    },
                    {
                      "color": "#13231B",
                      "fontSize": 16,
                      "lineHeight": 21,
                    },
                  ]
                }
                testID="category-title"
              >
                Cacti and Succulents
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
